### ISSUE TITLE
Facilitate the choice of BLAS provider (specifically MKL) at runtime

### ISSUE THREAD

**Issue by mdeff** [scipy/scipy#11812]
As conda-forge [successfully demonstrated](https://conda-forge.org/docs/maintainer/knowledge_base.html#switching-blas-implementation), it is possible to compile numpy and scipy against the netlib reference implementation of BLAS and LAPACK, and then switch to OpenBLAS, MKL, or BLIS at runtime.

To be able to switch to MKL without some functions segfaulting (#3193), one of the following should be done:
1. build scipy with `SCIPY_USE_G77_ABI_WRAPPER=1` (introduced in #10591 for conda-forge by @isuruf),
2. have `MKL_INTERFACE_LAYER=LP64,GNU` in the environment at runtime ([introduced by Intel](https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/499188) to fix #3193) ([also done in conda-forge](https://github.com/conda-forge/blas-feedstock/blob/master/recipe/libblas_mkl_activate.sh), just in case?).

To facilitate the work of distributors, it would make sense for the default build options to be maximally compatible. From #10591, "this wrapper works for all cases and incurs only a tiny overhead", which seems like a good tradeoff. Is there any other reason not to always use those wrappers? Or at least make it the default?

**Comment by pv** [comment]
The default build options correspond to BLAS that ABI-compatible with the Fortran compiler used.

The Fortran ABI is something that the binary distributor has to coordinate across all packages, and using ABI-incompatible compiler/library combo requires some sort of hack in all Fortran programs, so Scipy is not special. Most environments (Linux distributions, HPC environments, etc.) install libraries that are compatible with the Fortran compilers they are providing.

The MKL(g77)+gfortran combination in conda/conda-forge is a strange outlier with the ABI-incompatibility. AFAIK, MKL has also gfortran ABI-compatible libraries, but I'm not familiar with conda-forge so I don't know the reason why those are not used.

The wrappers are an ugly hack (incl. patching of vendored Fortran sources) that try to paper over ABI differences. Part of them uses cblas and is Fortran ABI-independent, but not everything is. E.g. the following Cython module:
```
from scipy.linalg.cython_lapack cimport cladiv
cdef float complex a = 1, b = 1
print("This should print (1+0j):")
print(cladiv(&a, &b))
```
prints garbage with wrong wrappers vs. ABI --- under conda-forge this works only for MKL with MKL_INTERFACE_LAYER env var unset (i.e. the case the wrappers are intended to work in). Arpack no longer uses these functions in versions several years ago, so the impact from these wrong results due to incompatible LAPACK is currently probably not critical in practice.

**Comment by isuruf** [comment]
> AFAIK, MKL has also gfortran ABI-compatible libraries, but I'm not familiar with conda-forge so I don't know the reason why those are not used.

Only on Linux. Not available on OSX

**Comment by pv** [comment]
I think the wrappers would be more cleanly implemented as a separate ABI-compatibility wrapper library, which would provide compatible symbols. The hurdle here is mainly distutils and getting the linker flags etc. right on all platforms. (For distributors, the distutils issue does not exist and there's better control over compilers etc. to get it right vs. platforms.)

This would allow us to stop patching sources, make it easier to provide 100% coverage of the relevant blas/lapack API instead of piecemeal, and avoid mistakes and the maintenance annoyance with the patching.

For Accelerate I remember seeing such existing (3rd-party) blas/lapack wrapper libraries --- maybe they exist also for MKL on OSX?

**Comment by mdeff** [comment]
Thanks for the detailed explanations @pv.

> AFAIK, MKL has also gfortran ABI-compatible libraries

So there's a third option: instead of symlinking `libblas.so` and `liblapack.so` to `libmkl_rt.so` and exporting `MKL_INTERFACE_LAYER=LP64,GNU`, we could symlink to another MKL library and avoid the export. @isuruf, do you know which library we should symlink to on Linux? (Maybe that's also what conda-forge should do to avoid the issue mentioned by @pv.)

Wouldn't a longer-term fix be to update cblas (upstream) to provide the full interface, and rely on cblas only?

**Comment by sturlamolden** [comment]
The «correct» solution to the ABI problem would be to use Fortran 2003 to write a C callable interface for each Fortran subroutine, and then export this as a DLL that is statically linked with MKL or OpenBLAS. The problem is you would have to do this for each Fortran subroutine in the exported C API (which presumably includes LAPACK as well as BLAS).

**Comment by pv** [comment]
The problem in this issue is that the Fortran compiler ABI is not compatible with MKL. Fortran 2003 won't help here, unless you use it to call the MKL C-API.

**Comment by isuruf** [comment]
@pv, thanks for the info. Looks like conda-forge setting doesn't work for `cladiv` and `zladiv`, although `cdotc, cdotu, zdotc, zdotu` works fine with MKL and OpenBLAS.
Looks like the wrapper can be improved here, https://github.com/scipy/scipy/blob/8dba340293fe20e62e173bdf2c10ae208286692f/scipy/_build_utils/src/wrap_g77_abi_f.f#L41
Any interest in fixing this? If so, I can send a PR.

**Comment by pv** [comment]
@isuruf: The problem is that cladiv/zladiv AFAIK don't exist in the LAPACK C-API, so you can't call them easily in an ABI-independent way. The wrapper code should be correct as is, but if you set `MKL_INTERFACE_LAYER=LP64,GNU` then MKL switches to gfortran ABI, and you'd need to have Scipy compiled without the wrappers.

Of course, it might be possible to look up MKL-dependent environment variables on the first call to cladiv/zladiv etc., but that gets ugly. However, that might be better as a conda-forge specific patch.

**Comment by isuruf** [comment]
How about vendoring the netlib code? It's just 2 lines. https://github.com/Reference-LAPACK/lapack/blob/9fb29d8ea47d643ad44a67d3da5de36e280143f5/SRC/cladiv.f#L89-L91

**Comment by pv** [comment]
That probably works.

But in long term, I would prefer to drop the wrappers altogether. For the OSX wrapper libraries I mentioned earlier: https://github.com/mcg1969/vecLibFort/ --- that could probably be adapted for MKL, and would solve the issues on a better level.

**Comment by isuruf** [comment]
@pv, one issue is that we want to support libraries linking directly to MKL as well. If the same process loads the wrapper library and MKL, there'll be symbol clashes right?

**Comment by isuruf** [comment]
https://github.com/scipy/scipy/pull/11863

**Comment by pv** [comment]
@isuruf: If you are able to swap `libblas -> openblas(gfortran ABI)` without things bombing, then I think a shim library as `libblas.so` won't be a problem either?

The details of how the shim libraries work I think depend on how the symbol resolution works out on the platform, and there are some differences between how it works in OSX and Linux/ELF.

**Comment by isuruf** [comment]
If a process loads both `libblas.so -> libmkl_rt.so` and `libmkl_rt.so`, then the symbol does the same thing. We don't recommend using `libblas.so -> libopenblas.so` and `libmkl_rt.so` both in the same process. It might work, it might not.

But `cladiv` from MKL and the wrapper library does two things, right?

(Yes, symbol resolution is complicated and I don't know enough about the differences)

**Comment by pv** [comment]
Right, I understood the main aim was to make libblas safe to swap for openblas-gfortran, so that all symbols in it should have gfortran ABI?

Mixing incompatible libs in same program I think is orthogonal to this (and I did not get earlier that this is your main concern). So yes, patching sources to avoid using incompatible symbols can solve that, and work for that is mostly done in Scipy.

However, from what I understand, the dynamic linking on osx has different namespace concept than Linux/ELF, and it might be worth to look if that can solve the name clash issues in a general way. There could also be a way to rename symbols as a postprocessing step on the built objects. These approaches could be better in the long run that patching Fortran sources.
